### PR TITLE
refactor interface for projections/proximal operators

### PR DESCRIPTION
**Pull request by Red-Portal**
This PR refactors how post-hoc modifications are applied to the iterates after performing a gradient descent step. For instance, before, updating the parameters of `LocationScale` always silently applied a projection step. Now, everything needs to be made into its own `OptimisationRule` to make it more modular and explicit.

More concretely, this PR changes the following:
- The scale matrix of a `LocationScale` distribution is no longer projected by default.
- A new keyword argument, `operator`, is added to `optimize`.
- The `operator` object is applied to the parameters after each gradient descent step.
- For the location-scale family, I added `ClipScale`, which clips the diagonal of the scale matrix to be strictly positive.

For example:
```julia
AdvancedVI.optimize(
    model,
    elbo,
    q_transformed,
    max_iter;
    adtype=ADTypes.AutoForwardDiff(),
    optimizer=Optimisers.Adam(1e-3),
    operator=ClipScale(),
)
```

For gradient descent, the operator is applied as:
```math
\lambda_{t+1} = \mathrm{operator}\left(\lambda_t + \gamma_t g_t\right),
```
where $g_t$ is the gradient estimator and $\gamma_t$ is the stepsize.

### CODE REVIEW

**Comment by Red-Portal** [comment]
@yebai I'll mark the v0.3 release (at last!) after this PR

**Review by mhauru — COMMENTED** [review]
Could we have a test where the eigenvalues drift too low and we can test both that it fails when nothing using `ProjectScale` and that it then succeeds when using `ProjectScale`? Just to very concretely see the effect, and see that the first case fails in the expected (rather than some other, unexpected) way.

Except for the above request, I'm happy with the software engineering. I would prefer it though if someone else who has views on the design choices here gave a second, approving opinion. I have little idea of what users need and want from their interfaces here, e.g. if the name `ProjectScale` is intuitive for users, or if there should be a warning if someone tries to optimise a `LocationScale` without using `ProjectScale`.

**Comment by Red-Portal** [comment]
Thank you both for chiming in!

@yebai I was thinking this to be similar in functionality to operations like [gradient clipping](https://fluxml.ai/Optimisers.jl/dev/api/#Optimisers.ClipNorm). How about I change the name to `ClipScale` and use the standard composition feature in `Optimisers.jl`, [`OptimiserChain`](https://fluxml.ai/Optimisers.jl/dev/api/#Optimisers.OptimiserChain)?

**Comment by Red-Portal** [comment]
@yebai Needed to change the API a little bit (see the summary at the top comment), do you agree with it?

File: docs/src/optimization.md [hunk 1]
```diff
@@ -26,3 +26,18 @@ PolynomialAveraging
 [^DCAMHV2020]: Dhaka, A. K., Catalina, A., Andersen, M. R., Magnusson, M., Huggins, J., & Vehtari, A. (2020). Robust, accurate stochastic optimization for variational inference. Advances in Neural Information Processing Systems, 33, 10961-10973.
 [^KMJ2024]: Khaled, A., Mishchenko, K., & Jin, C. (2023). Dowg unleashed: An efficient universal parameter-free gradient descent method. Advances in Neural Information Processing Systems, 36, 6748-6769.
 [^IHC2023]: Ivgi, M., Hinder, O., & Carmon, Y. (2023). Dog is sgd's best friend: A parameter-free dynamic step size schedule. In International Conference on Machine Learning (pp. 14465-14499). PMLR.
+## Operators
+
+Depending on the variational family, variational objective, and optimization strategy, it might be necessary to modify the variational parameters after performing a gradient-based update.
+For this, an operator acting on the parameters can be supplied via the  `operator` keyword argument of `optimize`.
```
[the remark this answers is not in the document]
Comment (Red-Portal) [reply docs/src/optimization.md:32]: Wouldn't `operate` be more explicit? I used `average` for the parameter averaging schemes, so I think it's more consistent. I was originally thinking about calling everything `apply` (for both the operators and the parameter averaging schemes). Then we are doing very different things only based on specialization, which I was thinking isn't very explicit.

File: src/AdvancedVI.jl [hunk 2]
```diff
@@ -242,6 +213,39 @@ include("optimization/averaging.jl")
 
 export NoAveraging, PolynomialAveraging
 
+# Operators for Optimization
+abstract type AbstractOperator end
+
+"""
+    operate(op, family, params, restructure)
```
[the remark this answers is not in the document]
Comment (Red-Portal) [reply src/AdvancedVI.jl:220]: Wouldn't operate be more explicit? I used average for the parameter averaging schemes, so I think it's more consistent. I was originally thinking about calling everything apply (for both the operators and the parameter averaging schemes). Then we are doing very different things only based on specialization, which I was thinking isn't very explicit.
